### ISSUE TITLE
PC Standalone App performs worse than when running on UnityEditor

### ISSUE THREAD

**Issue by gianluca-amprimo** [homuler/MediaPipeUnityPlugin#428]
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 20.04): Windows10
- Target Platform: Windows10
- MediaPipeUnityPlugin version or commit id: 0.7
- Bazel version:
- GCC/G++ version:

**Describe the current behavior**
I am using the system with a Kinect azure camera which is activated as a webcam stream. However handtracking works at very low fps (20) instead of 30 fps (camera framerate) when running in sync mode (I compute delta time for fps at each new output that is drawn). THIS BEHAVIOUR IS OBSERVED ONLY WHEN I BUILD THE EXECUTABLE. Inside Unity Editor it works at 30 fps

**Describe the expected behavior**
It should work at 30fps also outside the Unity Editor.

**Steps to reproduce the issue**
I have only added a fps checking in the sync if in handtrackingsolution. The issue is observed only when building the system as executable and running it outside editor. 

**Full logs**
[Player.log](https://github.com/homuler/MediaPipeUnityPlugin/files/7911251/Player.log)

**Additional context**
The system works perfectly fine inside the editor

**Comment by LGuilhermeRodrigues** [comment]
What's the resolution of your display? I was doing some tests here: when I'm using a 4K resolution the GPU reaches 100% of usage. When I change to FullHD there's no delay because GPU is only 50%.  Is that your case?
PS. I'm not using the new version of the plug-in

**Comment by gianluca-amprimo** [comment]
> 

I have a screen with 1280x720 resolution which is also the camera resolution.  I haven't observe any saturation problem, os resources demand is high but does not saturate resources either in Editor and outside. I am running the code on a minipc with Intel Core i5-10300H (quad-core 2.5 GHz, up to 4.5 GHz) and NVIDIA GeForce RTX 2060 6GB GDDR6 192-bit

**Comment by LGuilhermeRodrigues** [comment]
Ok, since it's not the same case, I'll investigate more about my problem and maybe open another issue.

**Comment by gianluca-amprimo** [comment]
@homuler could you provide me a feedback?

**Comment by homuler** [comment]
I'm not sure, but if the problem doesn't occur in UnityEditor, I feel it's a problem with the build settings on your side.

Please try or answer the below.

- Use the latest version (i.e. v0.8.2)
- Share the command you ran to build the library (e.g. `python build.py build --desktop cpu -vv`)
- Share the diff
   > I have only added a fps checking in the sync if in handtrackingsolution.
- If you've changed build settings, share the diff, too.
- If possible, find out which lines in particular take longer time

**Comment by gianluca-amprimo** [comment]
![log_editor_vs_player](https://user-images.githubusercontent.com/54496761/150755592-b69ac857-8212-4836-b120-57a067370eab.png)
![loader_message](https://user-images.githubusercontent.com/54496761/150755596-08093c47-0e7f-475c-9877-4a99b1e3f9e9.png)

@homuler thanks for your reply. While I try your suggestions, I have identified these differences in the log files produced by the run in the editor (Editor.log) and the run in build (Player.log). In Player.log there isn't the message about Tensorflow Lite delegate, is it possible that the delegate is missing? I have also noticed that it seems like some assets are not found, a c and a dll file...can you confirm me that this behaviour is normal?

**Comment by homuler** [comment]
> I have also noticed that it seems like some assets are not found, a c and a dll file

You can ignore these messages safely.

> In Player.log there isn't the message about Tensorflow Lite delegate, is it possible that the delegate is missing?

If XNNPACK delegate is not used, it can explain why it runs slower, but it looks like to me that the Glog logs are simply not being output to `Player.log`.

By the way, the sample app of the latest version is configured to be built in development mode by default.
I don't know which version you're using, so I'm not sure, but are you comparing performance in that state?

**Comment by gianluca-amprimo** [comment]
@homuler I have tried installing the lastest version and running it in debug mode, but the result is the same (the development build works at half the fps with respect to running the sample app in the Unity Editor. I have setup everything according to your Windows 10 tutorial.  Could it be something related to the fact that I am using Windows 10 Pro as operating system?

**Comment by gianluca-amprimo** [comment]
@homuler would it work if I add to build command you suggest in the tutorial the option  --define tflite_with_xnnpack=true to ensure that XNN PACK delegate is used in any case?

**Comment by homuler** [comment]
I just tested it, and XNNPACK delegate is being used properly, but I could confirm that the performance seems to be degraded even in my environment.

**Comment by gianluca-amprimo** [comment]
So where could this reduction in performance come from? I had a lot of D9002 and D9025 warning during build related to tflite and delegate, for example:

INFO: From Compiling tensorflow/lite/tflite_with_xnnpack_optional.cc:
cl : warning from command line D9002 : unknown option '-O3' will be ignored

which is also cited here https://github.com/tensorflow/tensorflow/issues/37744

and about D9025

INFO: From Compiling absl/base/internal/spinlock_wait.cc:
cl : warning from command line D9025 : override of '/w' with '/W3'

Are they normal? Could they be relevant? You find all of them in the txt file in my previous comment (sorry they are in italian but u can find them using the warning code). 
Thanks for ur replies, I am available for further investigation.

**Comment by homuler** [comment]
I did not find any significant difference in performance when measured using [Stopwatch](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.stopwatch?view=netstandard-2.1).
Can you confirm?
```diff
diff --git a/Assets/Mediapipe/Samples/Scenes/Hand Tracking/HandTrackingSolution.cs b/Assets/Mediapipe/Samples/Scenes/Hand Tracking/HandTrackingSolution.cs
index f1a78a1..4e6b2fa 100644
--- a/Assets/Mediapipe/Samples/Scenes/Hand Tracking/HandTrackingSolution.cs
+++ b/Assets/Mediapipe/Samples/Scenes/Hand Tracking/HandTrackingSolution.cs
@@ -8,6 +8,8 @@ using System.Collections;
 using System.Collections.Generic;
 using UnityEngine;

+using Stopwatch = System.Diagnostics.Stopwatch;
+
 namespace Mediapipe.Unity.HandTracking
 {
   public class HandTrackingSolution : Solution
@@ -21,6 +23,7 @@ namespace Mediapipe.Unity.HandTracking
     [SerializeField] private TextureFramePool _textureFramePool;

     private Coroutine _coroutine;
+    private Stopwatch _stopwatch;

     public RunningMode runningMode;

@@ -50,6 +53,7 @@ namespace Mediapipe.Unity.HandTracking
       }
       base.Play();
       _coroutine = StartCoroutine(Run());
+      _stopwatch = new Stopwatch();
     }

     public override void Pause()
@@ -123,6 +127,8 @@ namespace Mediapipe.Unity.HandTracking
       SetupAnnotationController(_handLandmarksAnnotationController, imageSource, true);
       SetupAnnotationController(_handRectsFromLandmarksAnnotationController, imageSource, true);

+      _stopwatch.Start();
+
       while (true)
       {
         yield return new WaitWhile(() => isPaused);
@@ -139,10 +145,12 @@ namespace Mediapipe.Unity.HandTracking
         if (runningMode == RunningMode.Sync)
         {
           // TODO: copy texture before `textureFrame` is released
+          _stopwatch.Restart();
           _screen.ReadSync(textureFrame);

           // When running synchronously, wait for the outputs here (blocks the main thread).
           var value = _graphRunner.FetchNextValue();
+          Debug.Log(_stopwatch.ElapsedMilliseconds);
           _palmDetectionsAnnotationController.DrawNow(value.palmDetections);
           _handRectsFromPalmDetectionsAnnotationController.DrawNow(value.handRectsFromPalmDetections);
           _handLandmarksAnnotationController.DrawNow(value.handLandmarks, value.handedness);
```

**Comment by gianluca-amprimo** [comment]
Yes, I can confirm the behaviour...

**Comment by gianluca-amprimo** [comment]
I see what u meant, but you are only evaluating inference time, whereas the way in which I measured time was including the full time from graph output to next graph output (including point visualization on screen) so the two approach cannot really be compared. Hence, could be the issue related not to inference time but to the full visualization procedure (I still experience the problem).

**Comment by homuler** [comment]
I see. Please try this.
```diff
diff --git a/Assets/Mediapipe/Samples/Common/Scripts/WaitForResult.cs b/Assets/Mediapipe/Samples/Common/Scripts/WaitForResult.cs
index 6fbb3b2..ea31325 100644
--- a/Assets/Mediapipe/Samples/Common/Scripts/WaitForResult.cs
+++ b/Assets/Mediapipe/Samples/Common/Scripts/WaitForResult.cs
@@ -53,6 +53,14 @@ namespace Mediapipe.Unity
             break;
           }
           tmpResult = _inner.Current;
+
+          if (_inner.Current is CustomYieldInstruction customYieldInstruction)
+          {
+            if (!customYieldInstruction.keepWaiting)
+            {
+              continue;
+            }
+          }
         }
         catch (Exception e)
         {
diff --git a/Assets/Mediapipe/Samples/Scenes/Hand Tracking/HandTrackingSolution.cs b/Assets/Mediapipe/Samples/Scenes/Hand Tracking/HandTrackingSolution.cs
index f1a78a1..69371a7 100644
--- a/Assets/Mediapipe/Samples/Scenes/Hand Tracking/HandTrackingSolution.cs
+++ b/Assets/Mediapipe/Samples/Scenes/Hand Tracking/HandTrackingSolution.cs
@@ -8,6 +8,8 @@ using System.Collections;
 using System.Collections.Generic;
 using UnityEngine;

+using Stopwatch = System.Diagnostics.Stopwatch;
+
 namespace Mediapipe.Unity.HandTracking
 {
   public class HandTrackingSolution : Solution
@@ -21,6 +23,7 @@ namespace Mediapipe.Unity.HandTracking
     [SerializeField] private TextureFramePool _textureFramePool;

     private Coroutine _coroutine;
+    private Stopwatch _stopwatch;

     public RunningMode runningMode;

@@ -50,6 +53,7 @@ namespace Mediapipe.Unity.HandTracking
       }
       base.Play();
       _coroutine = StartCoroutine(Run());
+      _stopwatch = new Stopwatch();
     }

     public override void Pause()
@@ -123,18 +127,25 @@ namespace Mediapipe.Unity.HandTracking
       SetupAnnotationController(_handLandmarksAnnotationController, imageSource, true);
       SetupAnnotationController(_handRectsFromLandmarksAnnotationController, imageSource, true);

+      long a, b, c, d, e, f;
+
       while (true)
       {
+        _stopwatch.Restart();
         yield return new WaitWhile(() => isPaused);

         var textureFrameRequest = _textureFramePool.WaitForNextTextureFrame();
         yield return textureFrameRequest;
         var textureFrame = textureFrameRequest.result;
+        a = _stopwatch.ElapsedMilliseconds;

         // Copy current image to TextureFrame
         ReadFromImageSource(imageSource, textureFrame);
+        b = _stopwatch.ElapsedMilliseconds;

         _graphRunner.AddTextureFrameToInputStream(textureFrame).AssertOk();
+        c = _stopwatch.ElapsedMilliseconds;
+        d = e = c;

         if (runningMode == RunningMode.Sync)
         {
@@ -143,13 +154,18 @@ namespace Mediapipe.Unity.HandTracking

           // When running synchronously, wait for the outputs here (blocks the main thread).
           var value = _graphRunner.FetchNextValue();
+          d = _stopwatch.ElapsedMilliseconds;
           _palmDetectionsAnnotationController.DrawNow(value.palmDetections);
           _handRectsFromPalmDetectionsAnnotationController.DrawNow(value.handRectsFromPalmDetections);
           _handLandmarksAnnotationController.DrawNow(value.handLandmarks, value.handedness);
           _handRectsFromLandmarksAnnotationController.DrawNow(value.handRectsFromLandmarks);
+          e = _stopwatch.ElapsedMilliseconds;
         }

         yield return new WaitForEndOfFrame();
+        f = _stopwatch.ElapsedMilliseconds;
+
+        Debug.Log($"FrameRequest: {a}, Copy: {b - a}, Send: {c - b}, Fetch: {d - c}, Render: {e - d}, Yield 2: {f - e}");
       }
     }
 
```

When the app runs at 60FPS, it takes about 16ms to wait for one frame, which seems to cause this issue.

**Comment by gianluca-amprimo** [comment]
I tried the code you provided me. There is  quite a huge gap in the time for the frame request (2/3 ms vs >9 ms). Where do you think this issue could come from?

**Comment by homuler** [comment]
So can we think the above code solve this issue?

> Where do you think this issue could come from?

I wrote that in my comment above.
> When the app runs at 60FPS, it takes about 16ms to wait for one frame, which seems to cause this issue.

**Comment by gianluca-amprimo** [comment]
The gap between the two versions (editor and build) is still not completely closed though even using your code (I suppose the modification you are talking about is the one in WaitForResult.cs)...Why is it not possible to have the same frame request time in both the versions?

**Comment by homuler** [comment]
I see, so you were talking about the time gap between running on UnityEditor and Standalone.

You'll have to ask Unity developers why it runs so fast in UnityEditor, but it all boils down to this comment.
> When the app runs at 60FPS, it takes about 16ms to wait for one frame, which seems to cause this issue.

Try this sample code if you're interested (I tested in an (almost) empty project).
```cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

using Stopwatch = System.Diagnostics.Stopwatch;

public class CoroutineTest : MonoBehaviour
{
  // Start is called before the first frame update
  void Start()
  {
    StartCoroutine(Run());
  }

  private IEnumerator Run()
  {
    var stopwatch = new Stopwatch();

    while (true)
    {
      long a, b, c;
      stopwatch.Restart();

      yield return new WaitWhile(() => false);
      a = stopwatch.ElapsedMilliseconds;

      yield return new WaitUntil(() => true);
      b = stopwatch.ElapsedMilliseconds;

      yield return new WaitForEndOfFrame();
      c = stopwatch.ElapsedMilliseconds;

      Debug.Log($"WaitWhile: {a}, WaitUntil: {b - a}, WaitForEndOfFrame: {c - b}");
    }
  }
}
```

Here are excerpts of the logs when I ran it in my environment.
- Editor.log
   ```txt
   ...
   WaitWhile: 0, WaitUntil: 0, WaitForEndOfFrame: 1
   WaitWhile: 0, WaitUntil: 0, WaitForEndOfFrame: 1
   WaitWhile: 0, WaitUntil: 1, WaitForEndOfFrame: 0
   WaitWhile: 0, WaitUntil: 0, WaitForEndOfFrame: 1
   ...
   ```

- Player.log
   ```txt
   ...
   WaitWhile: 16, WaitUntil: 17, WaitForEndOfFrame: 0
   WaitWhile: 16, WaitUntil: 17, WaitForEndOfFrame: 0
   WaitWhile: 16, WaitUntil: 16, WaitForEndOfFrame: 1
   WaitWhile: 16, WaitUntil: 17, WaitForEndOfFrame: 0
   ...
   ```

After all, I guess we should avoid to yield `IEnumerator` if not necessary.

**Comment by gianluca-amprimo** [comment]
Yes, I confirm that the issue is related to those yeld instruction.

**Comment by ROBYER1** [comment]
We are finding the same issue here, in Editor the responsiveness is much faster than the build. Is there anything we can do to help solve this?

**Comment by ROBYER1** [comment]
Thanks for your hard work on this @homuler 🧡

**Comment by homuler** [comment]
I think this issue is resolved by #471 and #472 because now the `yield` instruction is called only once in each loop when running in synchronous mode.

However, as mentioned in https://github.com/homuler/MediaPipeUnityPlugin/issues/428#issuecomment-1026435131, if the built app runs at 60FPS, each sample scene may run faster on UnityEditor.
If the app can run in your environment at the same FPS as in UnityEditor, the new Non-Blocking sync mode should run faster than the synchronous mode.

If the issue is not fixed, please let me know.

**Comment by ROBYER1** [comment]
Just to add to this, if anyone is confused why their app runs worse in build than editor still, if you goto Project Settings > Quality and for any quality settings you have including default, set VSync Count to 'Don't Sync' so that your build can run faster than the refresh rate of the monitor you are using.
